### ISSUE TITLE
Numerous errors trying to build with MSYS2 MINGW64 GCC

### ISSUE THREAD

**Issue by nimdam** [amarao/flashnul#3]
Hi.

My gcc 4.9.2 build is MINGW64 from MSYS2.
Tried to compile flashnul myself, mainly to test building software from sources. I am newbie here, so faced with troubles i couldn't fix myself. Any help is appreciated. Here are my steps from the very beginning.

`$ git clone https://github.com/amarao/flashnul.git`
`$ cd flashnul/src/`
`$ make all`

>cc --pipe -O2 -march=i686 -Wall   -c -o flashnul.o flashnul.c
flashnul.c:1:0: error: CPU you selected does not support x86-64 instruction set
 #include <stdarg.h>
 ^
flashnul.c:1:0: error: CPU you selected does not support x86-64 instruction set
<builtin>: recipe for target 'flashnul.o' failed
make: *** [flashnul.o] Error 1

`$ make all CC="gcc -m32"`

>gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o flashnul.o flashnul.c
gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o generate.o generate.c
gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o io.o io.c
In file included from io.c:10:0:
ddk.h:7:14: error: nested redefinition of 'enum _STORAGE_QUERY_TYPE'
 typedef enum _STORAGE_QUERY_TYPE {
              ^
ddk.h:7:14: error: redeclaration of 'enum _STORAGE_QUERY_TYPE'
In file included from G:/msys64/mingw64/x86_64-w64-mingw32/include/winscard.h:11:0,
                 from G:/msys64/mingw64/x86_64-w64-mingw32/include/windows.h:97,
                 from io.c:1:
G:/msys64/mingw64/x86_64-w64-mingw32/include/winioctl.h:1740:14: note: originally defined here
 typedef enum _STORAGE_QUERY_TYPE {
              ^
In file included from io.c:10:0:
ddk.h:8:5: error: redeclaration of enumerator 'PropertyStandardQuery'
     PropertyStandardQuery = 0,          // Retrieves the descriptor
     ^
In file included from G:/msys64/mingw64/x86_64-w64-mingw32/include/winscard.h:11:0,
                 from G:/msys64/mingw64/x86_64-w64-mingw32/include/windows.h:97,
                 from io.c:1:
G:/msys64/mingw64/x86_64-w64-mingw32/include/winioctl.h:1741:3: note: previous definition of 'PropertyStandardQuery' was here
   PropertyStandardQuery     = 0,
   ^

...

`here goes numerous identical redefinition and redeclaration errors`

...

><builtin>: recipe for target 'io.o' failed
make: *** [io.o] Error 1

Google shows plenty of redefinition and redeclaration errors fixes. As i understood ddk.h declares and defines C structures already defined in x86_64-w64-mingw32/include/xxxx.h. So commented inclusion string (//#include "ddk.h") in io.c, made "make clean" and tried again. Managed to go further, but no luck again.

`$ make all CC="gcc -m32"`

>gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o flashnul.o flashnul.c
gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o generate.o generate.c
gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o io.o io.c
io.c: In function 'get_adapter_property':
io.c:436:3: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   dev->bus_type = ( (PSTORAGE_ADAPTER_DESCRIPTOR)buffer )->BusType;
   ^
io.c: In function 'get_device_property':
io.c:475:26: warning: pointer targets in passing argument 1 of 'get_name' differ in signedness [-Wpointer-sign]
   dev->vendor = get_name(buffer,device->VendorIdOffset,returned_length);
                          ^
io.c:351:7: note: expected 'const char *' but argument is of type 'unsigned char *'
 char* get_name (const char* buffer, int text_offset,int max_len){
       ^
io.c:476:24: warning: pointer targets in passing argument 1 of 'get_name' differ in signedness [-Wpointer-sign]
   dev->name = get_name(buffer,device->ProductIdOffset,returned_length);
                        ^
io.c:351:7: note: expected 'const char *' but argument is of type 'unsigned char *'
 char* get_name (const char* buffer, int text_offset,int max_len){
       ^
io.c:477:28: warning: pointer targets in passing argument 1 of 'get_name' differ in signedness [-Wpointer-sign]
   dev->revision = get_name(buffer,device->ProductRevisionOffset,returned_length);
                            ^
io.c:351:7: note: expected 'const char *' but argument is of type 'unsigned char *'
 char* get_name (const char* buffer, int text_offset,int max_len){
       ^
io.c:478:30: warning: pointer targets in passing argument 1 of 'get_name' differ in signedness [-Wpointer-sign]
   dev->raw_serial = get_name(buffer,device->SerialNumberOffset,returned_length);
                              ^
io.c:351:7: note: expected 'const char *' but argument is of type 'unsigned char *'
 char* get_name (const char* buffer, int text_offset,int max_len){
       ^
gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o display.o display.c
gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o actions.o actions.c
actions.c: In function 'quick_test':
actions.c:467:18: warning: variable 'pattern' set but not used [-Wunused-but-set-variable]
  unsigned char*  pattern   = NULL;
                  ^
gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o common.o common.c
gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o detect.o detect.c
detect.c: In function 'list_logical_drives':
detect.c:65:7: warning: suggest parentheses around operand of '!' or change '&' to '&&' or '!' to '~' [-Wparentheses]
   if( !flags & 0x0100 )
       ^
detect.c:45:9: warning: variable 'res_space' set but not used [-Wunused-but-set-variable]
  DWORD  res_space;
         ^
gcc -m32 --pipe -O2 -march=i686 -Wall   -c -o advanced.o advanced.c
make -C colian
make[1]: Entering directory «/home/nimdam/flashnul/src/colian»
gcc -m32 -O3 --pipe    -c -o base.o base.c
gcc -m32 -O3 --pipe    -c -o cmd.o cmd.c
gcc -m32 -O3 --pipe    -c -o dbg.o dbg.c
gcc -m32 -O3 --pipe    -c -o group.o group.c
gcc -m32 -O3 --pipe    -c -o init-close.o init-close.c
gcc -m32 -O3 --pipe    -c -o join.o join.c
gcc -m32 -O3 --pipe    -c -o print.o print.c
gcc -m32 -O3 --pipe    -c -o read.o read.c
gcc -m32 -O3 --pipe    -c -o write.o write.c
gcc -m32 -O3 --pipe    -c -o env.o env.c
ar cr libcolian.a base.o cmd.o dbg.o  group.o init-close.o join.o print.o read.o write.o env.o
ranlib libcolian.a
make[1]: Leaving directory «/home/nimdam/flashnul/src/colian»
gcc flashnul.o generate.o io.o display.o actions.o common.o detect.o advanced.o   -L colian -lcolian  -lkernel32 -o flashnul.exe
G:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: skipping incompatible colian/libcolian.a when searching for -lcolian
G:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: skipping incompatible colian\libcolian.a when searching for -lcolian
G:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: skipping incompatible colian/libcolian.a when searching for -lcolian
G:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lcolian
collect2.exe: error: ld returned 1 exit status
Makefile:7: recipe for target 'all' failed
make: *** [all] Error 1

Ok. Fixing colian\Makefile first line: CC=gcc -m32. Let's try one more time.
The same errors again. Something with ld.exe linker, seems to be x64 problem. Have no ideas currently. May be MINGW64 GCC issue, i'll try to make some more testing.

**Comment by nimdam** [comment]
By the way, i've tried to remove "-march=i686" from Makefile and got this:

`$ make all`

>cc --pipe -O2 -Wall   -c -o flashnul.o flashnul.c
flashnul.c: In function ‘drive_number’:
flashnul.c:160:3: warning: array subscript has type ‘char’ [-Wchar-subscripts]
   if(!isdigit(name[c])){
   ^
flashnul.c: In function ‘isblocked’:
flashnul.c:200:5: warning: array subscript has type ‘char’ [-Wchar-subscripts]
     if(strchr(soptM(config,OPT_block_letter,i),tolower(letter)))
     ^
flashnul.c:202:5: warning: array subscript has type ‘char’ [-Wchar-subscripts]
     if(strchr(soptM(config,OPT_block_letter,i),toupper(letter)))
     ^
flashnul.c: In function ‘prompt’:
flashnul.c:233:2: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘long long int’ [-Wformat=]
  printf("\tSelected drive:\t\t %s, %I64db (%s)\n\n", dev->user_name, dev->size, human_view( dev->size, 'b') );
  ^
flashnul.c:233:2: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘long long int’ [-Wformat=]
flashnul.c:245:4: warning: array subscript has type ‘char’ [-Wchar-subscripts]
    reply[c] = toupper( reply[c] );
    ^
flashnul.c: In function ‘read_config_file’:
flashnul.c:283:26: error: ‘_pgmptr’ undeclared (first use in this function)
    char *home_dir=strdup(_pgmptr);
                          ^
flashnul.c:283:26: note: each undeclared identifier is reported only once for each function it appears in
<builtin>: recipe for target 'flashnul.o' failed
make: *** [flashnul.o] Error 1

**Comment by mingwandroid** [comment]
It appears that you are trying to compile some really old, really broken code. I'm not going to help you with this as it's got *nothing* to do with MSYS2/MinGW-w64 and I'm not interested in it personally.

**Comment by amarao** [comment]
Sorry, guys, but I will not fix issues with x86_64.  Program has been created in antique time for antique operating system which I used 6-8 years ago.

I'm closing this issue for now. If you got any pull requests with fixes, feel free to create new issue.
